# Racon
Consensus module for raw de novo DNA assembly of long uncorrected reads.  

## Description
Racon is intended as a standalone consensus module to correct raw contigs generated by rapid assembly methods which do not include a consensus step, such as [Miniasm](https://github.com/lh3/miniasm).  
The goal of Racon is to generate genomic consensus which is of similar or better quality compared to the output generated by assembly methods which employ both error correction and consensus steps, while providing a speedup of several times compared to those methods.  
Racon takes as input only two files: the raw contig FASTA and the alignments (SAM format) of raw reads mapped to the same contigs.  
We recommend [GraphMap](https://github.com/isovic/graphmap) for generating the alignments.  

Please note: Racon depends on *quality values* - the SAM file needs to have them included.  
There is also a default QV threshold set to 10 (Phred score). If your data is of poorer quality, you can modify this parameter using ```--qv FLOAT```. If ```FLOAT``` is ```-1```, the QV filtering will be turned off.  

## Quick start
Clone and make Racon:
```  
git clone https://github.com/isovic/racon.git  && cd racon && make modules && make tools && make -j  
```
Run an example script:  
```  
./example1-lambda.sh  
```  
Tip: Running Racon iteratively will produce better consensus sequences. (But don't forget to align your reads to the consensus sequence from the previous iteration.)  

## Dependencies
1. gcc >= 4.8  

## Installation
```  
git clone https://github.com/isovic/racon.git  
cd racon  
make modules  
make tools  
make -j  
```  

## Usage
```  
bin/racon [options] <raw_contigs.fasta> <alignments.sam> <out_consensus.fasta>  
```  
Racon depends on quality values - the SAM file needs to have them included.  
For detailed info on various options, run ```bin/racon``` without arguments.  

## Contact information

For additional information, help and bug reports please send an email to one of the following: robert.vaser@fer.hr,
ivan.sovic@irb.hr, mile.sikic@fer.hr


## Acknowledgment

This work has been supported in part by Croatian Science Foundation under the project UIP-11-2013-7353.
