# Racon

[![Published in Genome Research](https://img.shields.io/badge/published%20in-Genome%20Research-blue.svg)](https://doi.org/10.1101/gr.214270.116)

Consensus module for raw de novo DNA assembly of long uncorrected reads.

## Description
Racon is intended as a standalone consensus module to correct raw contigs generated by rapid assembly methods which do not include a consensus step, such as [Miniasm](https://github.com/lh3/miniasm).  
The goal of Racon is to generate genomic consensus which is of similar or better quality compared to the output generated by assembly methods which employ both error correction and consensus steps, while providing a speedup of several times compared to those methods.  
Racon takes as input only three files: contigs in FASTA/FASTQ format, reads in FASTQ format and overlaps/alignments between the reads and the contigs in MHAP/PAF/SAM format. Output is a set of polished contigs in FASTA format printed to stdout.

Racon can as well be used as a read error-correction tool. In this scenario, the MHAP/PAF/SAM file needs to contain actual overlaps instead of mappings without any dual or self overlaps.  

## Dependencies
1. gcc 4.8+ or clang 3.4+
2. cmake 3.2+

## Installation
To install Racon run the following commands:

```bash
    git clone --recursive https://github.com/isovic/racon.git racon
    cd racon
    mkdir build
    cd build
    cmake -DCMAKE_BUILD_TYPE=Release ..
    make
```

After successfull installation, an executable named `racon` will appear in `build/bin`.

Optionally, you can run `sudo make install` to install racon executable to your machine.

***Note***: if you omitted `--recursive` from `git clone`, run `git submodule init` and `git submodule update` before proceeding with compilation.

## Usage
Usage of racon is as following:

    racon [options ...] <sequences> <overlaps> <target sequences>

        <sequences>
            input file in FASTA/FASTQ format containing seqeunces used for
            correction
        <overlaps>
            input file in MHAP/PAF/SAM format containing overlaps between
            sequences and target sequences
        <target seqeunces>
            input file in FASTA/FASTQ format containing sequences which will
            be corrected

    options:
        -u, --include-unpolished
            output unpolished target sequences (each sequence contains a tag in
            its header (C:<float>) which represents the percentage of polished
            windows)
        -f, --fragment-correction
            perform fragment correction instead of contig polishing (overlaps
            file should not contain dual/self overlaps!)
        -w, --window-length <int>
            default: 500
            size of window on which POA is performed
        -m, --match <int>
            default: 5
            score for matching bases
        -x, --mismatch <int>
            default: -4
            score for mismatching bases
        -g, --gap <int>
            default: -8
            gap penalty (must be negative)
        -t, --threads <int>
            default: 1
            number of threads
        -h, --help
            prints out the help

## Contact information

For additional information, help and bug reports please send an email to one of the following: ivan.sovic@irb.hr, robert.vaser@fer.hr, mile.sikic@fer.hr, nagarajann@gis.a-star.edu.sg

## Acknowledgment

This work has been supported in part by Croatian Science Foundation under the project UIP-11-2013-7353. IS is supported in part by the Croatian Academy of Sciences and Arts under the project "Methods for alignment and assembly of DNA sequences using nanopore sequencing data". NN is supported by funding from A*STAR, Singapore.
